File: Packages/NoteroMidiPlugin/Runtime/Windows/MidiDeviceAdapter/PopPiano.cs
#if !UNITY_ANDROID || UNITY_EDITOR
using NAudio.Midi;
#endif

#if !UNITY_ANDROID || UNITY_EDITOR
namespace Notero.MidiPlugin.Windows.MidiDeviceAdapter
{
    public class PopPianoMidiDeviceAdapter : MidiDevice
    {
        private MidiOut m_MidiOut;

        private PopPianoMidiDeviceAdapter(MidiOutCapabilities midiOutCapabilities, MidiOut midiOut) : base(midiOutCapabilities, midiOut)
        {
            this.m_MidiOut = midiOut;
        }

        public static DeviceAssertion DeviceAssertion => Assert;

        private static MidiDevice Assert(MidiOutCapabilities midiDeviceInfo, MidiOut midiOut)
        {
            var deviceName = midiDeviceInfo.ProductName;

            return deviceName.ToLower().StartsWith("pop piano") ? new PopPianoMidiDeviceAdapter(midiDeviceInfo, midiOut) : null;
        }

        public override void SendLEDControlMessage(bool isOn, int keyIndex)
        {
            var stateByte = isOn ? 1 : 0;
            var buffer = new byte[8];

            buffer[0] = 0xF0;
            buffer[1] = 0x4D;
            buffer[2] = 0x4C;
            buffer[3] = 0x4E;
            buffer[4] = 0x45;
            buffer[5] = (byte)keyIndex;
            buffer[6] = (byte)stateByte;
            buffer[7] = 0xF7;

            m_MidiOut.SendBuffer(buffer);
        }

        public override void SendVolumeControlMessage(int volumeAsPercent)
        {
            var buffer = new byte[5];
            var asDecimal = (16 / 100f * volumeAsPercent) * 7.5625f;

            buffer[0] = 0xF0;
            buffer[1] = 0x5F;
            buffer[2] = 0x70;
            buffer[2] = (byte)asDecimal;
            buffer[2] = 0xF7;

            m_MidiOut.SendBuffer(buffer);
        }

        public override void SendMidiEvent(byte[] buffer)
        {
            m_MidiOut.SendBuffer(buffer);
        }
    }
}
#endif

File: Packages/NoteroMidiPlugin/Runtime/Native/midiDeviceAdapter/PopPiano.kt
package co.notero.midiplugin.midiDeviceAdapter

import android.media.midi.MidiDeviceInfo
import android.util.Log
import co.notero.midiplugin.NoteroMidiDevice
import co.notero.midiplugin.NoteroMidiInputPort
import co.notero.midiplugin.TAG
import kotlin.math.roundToInt

class PopPianoMidiDeviceAdapter (
    target: MidiDeviceInfo,
    private val inputPort: NoteroMidiInputPort
) : NoteroMidiDevice(target, inputPort) {
    companion object {
        val deviceAssertion: DeviceAssertion = { midiDeviceInfo: MidiDeviceInfo, inputPort: NoteroMidiInputPort -> assert(midiDeviceInfo,inputPort) }
        fun assert(
            midiDeviceInfo: MidiDeviceInfo,
            inputPort: NoteroMidiInputPort,
        ): NoteroMidiDevice? {

            // need to use `toLowerCase` due to unsupported `lowercase` in unity
            if (midiDeviceInfo.properties.getString("name")?.toLowerCase()?.trim() == "pop piano") {
                val buffer = ByteArray(15)
                Thread.sleep(300) // need explanation here why??
                buffer[0] = 0xF0.toByte()
                buffer[1] = 0x0F.toByte()
                buffer[2] = 0x33.toByte()
                buffer[3] = 0x16.toByte()
                buffer[4] = 0x0.toByte()
                buffer[5] = 0x0.toByte()
                buffer[6] = 0x0.toByte()
                buffer[7] = 0x0.toByte()
                buffer[8] = 0x0.toByte()
                buffer[9] = 0x0.toByte()
                buffer[10] = 0x0.toByte()
                buffer[11] = 0x0.toByte()
                buffer[12] = 0x0.toByte()
                buffer[13] = 0x0.toByte()
                buffer[14] = 0xF7.toByte()
                inputPort.send(buffer, 0, 15)
                val p = PopPianoMidiDeviceAdapter(midiDeviceInfo, inputPort)
                return p
            }
            return null;
        }


    }

    override fun sendLEDControlMessage(isOn: Boolean, keyIndex: Int) {
        Log.d(TAG, "Send Key ${if (isOn) "On" else "Off"} for key ${keyIndex.toByte()}")
        val stateByte = if (isOn) 17 else 0x00
        val messageLen = 8
        val buffer = ByteArray(messageLen)
        buffer[0] = 0xF0.toByte()
        buffer[1] = 0x4D.toByte()
        buffer[2] = 0x4C.toByte()
        buffer[3] = 0x4E.toByte()
        buffer[4] = 0x45.toByte()
        buffer[5] = keyIndex.toByte()
        buffer[6] = stateByte.toByte()
        buffer[7] = 128.toByte()
        buffer[7] = 0xF7.toByte()
        inputPort.send(buffer, 0, messageLen)
    }

    override fun sendVolumeControlMessage(volumeAsPercent: Int) {
        val messageLen = 5
        val buffer = ByteArray(messageLen)
        val asDecimal = (16 / 100f * volumeAsPercent) * 7.5625f;
        buffer[0] = 0xF0.toByte()
        buffer[1] = 0x5F.toByte()
        buffer[2] = 0x70.toByte()
        buffer[3] = asDecimal.roundToInt().toByte()
        buffer[4] = 0xF7.toByte()
        inputPort.send(buffer, 0, messageLen)
    }

    override fun sendMidiEvent(buffer: ByteArray) {
        inputPort.send(buffer, 0, buffer.size)
    }

}
